gbasis/evals/density.py

```diff
@@ -95,12 +95,170 @@ def evaluate_density(one_density_matrix, basis, points, transform=None, threshol
-    output = evaluate_density_using_evaluated_orbs(one_density_matrix, orb_eval)
+    density = evaluate_density_using_evaluated_orbs(one_density_matrix, orb_eval)
     # Fix #117: check magnitude of small negative density values, then use clip to remove them
-    min_output = np.min(output)
-    if min_output < 0.0 and abs(min_output) > threshold:
-        raise ValueError(f"Found negative density <= {-threshold}, got {min_output}.")
-    return output.clip(min=0.0)
+    min_density = np.min(density)
+    if min_density < 0.0 and abs(min_density) > threshold:
+        raise ValueError(f"Found negative density <= {-threshold}, got {min_density}.")
+    return density.clip(min=0.0)
+
+
+def evaluate_dm_using_evaluated_orbs(one_density_matrix, orb_eval_list):
+    """Return the evaluation of the density matrix given the evaluated orbitals.
+
+    Parameters
+    ----------
+    one_density_matrix : np.ndarray(K_orb, K_orb)
+        One-electron density matrix in terms of the given orbitals.
+    orb_eval_list : list of orbitals evaluated on grid points [np.ndarray(K_orb, N) ...]
+        Orbitals evaluated at :math:`N` grid points.
+        The set of orbitals must be the same basis set used to build the one-electron density
+        matrix.
+
+    Returns
+    -------
+    density : np.ndarray(N1,N2,K_orb,K_orb)
+        Density Matrix evaluated at `N1,N2` grid points.
+
+    Raises
+    ------
+    TypeError
+        If `orb_eval_list` does not consist of 2-dimensional `numpy` arrays with `dtype` float.
+        If `one_density_matrix` is not a symmetric, 2-dimensional `numpy` array with `dtype` float.
+
+    """
+    if not (
+        isinstance(one_density_matrix, np.ndarray)
+        and one_density_matrix.ndim == 2
+        and one_density_matrix.dtype == float
+    ):
+        raise TypeError(
+            "One-electron density matrix must be a two-dimensional `numpy` array with `dtype`"
+            " float."
+        )
+    for orb in orb_eval_list:
+        if not (isinstance(orb, np.ndarray) and orb.ndim == 2 and orb.dtype == float):
+            raise TypeError(
+                "Evaluation of orbitals must be a two-dimensional `numpy` array with `dtype` float."
+            )
+    if one_density_matrix.shape[0] != one_density_matrix.shape[1]:
+        raise ValueError("One-electron density matrix must be a square matrix.")
+
+    if not np.allclose(one_density_matrix, one_density_matrix.T):
+        raise ValueError("One-electron density matrix must be symmetric.")
+
+    # Tensor product for \gamma(\mathbf{r}_1,\mathbf{r}_2) = \sum_{pq} \gamma_{pq} \chi_p(\mathbf{r}_1) \chi_q(\mathbf{r}_2)
+    density = np.einsum("ij,ik,jl->klij", one_density_matrix, orb_eval_list[0], orb_eval_list[1])
+
+    # returns dm evaluated on each grid point
+    return density
+
+
+def evaluate_dm_density(one_density_matrix, basis, points_list, transform=None):
+    r"""Return the density of the given basis set at the given points.
```

Comment:
If I checked correctly this is the same docstring as in `evaluate_density`. I know they are similar but here two sets of points are used instead of one. Maybe we can add this in the docstring.